core/app/models/spree/promotion_action.rb

```diff
@@ -19,5 +19,15 @@ class PromotionAction < Spree::Base
     def perform(options = {})
       raise 'perform should be implemented in a sub-class of PromotionAction'
     end
+
+    def duplicate
+      duplicated_action = self.dup
+      duplicate_calculator(duplicated_action) if try(:calculator)
+      duplicated_action
+    end
+
+    def duplicate_calculator(dupe)
```

Comment:
Not sure if we gain a lot from this method since it's a one liner and the following is a bit clearer and you don't have to pass around a variable. At the very least the `dupe` variable should be `action` or something similar since that's the only thing it can be.

```
dulication.action.calculator = calculator.dup if calculator 
```